### ISSUE TITLE
Client Customer Id is required if CustomerIdMode is set to CLIENT_EXTERNAL_CUSTOMER_ID.

### ISSUE THREAD

**Issue by tonmanna** [googleads/googleads-dotnet-lib#160]
I'm trying to upgrade to the last version sometimes we found exception about 
Client Customer Id is required if CustomerIdMode is set to CLIENT_EXTERNAL_CUSTOMER_ID.

When rollback to 22.20 not found any problem. I'm trying to investigate this problem but unfortunately is not always happened.

**Comment by pieceofsummer** [comment]
Was working fine on 23.1.1 until yesterday, now getting the same issue.

It appears that they've changed something on the server, so now it always expects headers (clientCustomerId, developerToken etc.) in namespace `https://adwords.google.com/api/adwords/cm/{version}` instead of service namespace, so it stopped working for services from other namespaces (e.g. `https://adwords.google.com/api/adwords/billing/{version}`).

Temporarily fixed it by implementing a custom header class:
```c#
public class FixedRequestHeader : RequestHeader
{
    private readonly string _namespace;
    private readonly string _elementNamespace;

    public FixedRequestHeader(RequestHeader header)
    {
        _namespace = header.Namespace;
        _elementNamespace = "https://adwords.google.com/api/adwords/cm/" + header.Version;

        clientCustomerId = header.clientCustomerId;
        developerToken = header.developerToken;
        userAgent = header.userAgent;
        partialFailure = header.partialFailure;
        partialFailureSpecified = header.partialFailureSpecified;
        validateOnly = header.validateOnly;
        validateOnlySpecified = header.validateOnlySpecified;
    }

    public override string Namespace => _namespace;

    protected override void OnWriteHeaderContents(XmlDictionaryWriter writer, MessageVersion messageVersion)
    {
        if (!string.IsNullOrEmpty(clientCustomerId))
            writer.WriteElementString("clientCustomerId", _elementNamespace, clientCustomerId);

        writer.WriteElementString("developerToken", _elementNamespace, developerToken);
        writer.WriteElementString("userAgent", _elementNamespace, userAgent);

        if (validateOnlySpecified)
            writer.WriteElementString("validateOnly", _elementNamespace, validateOnly.ToString().ToLower());

        if (partialFailureSpecified)
            writer.WriteElementString("partialFailure", _elementNamespace, partialFailure.ToString().ToLower());
    }
}
```

and replacing default header when calling the service:
```c#
var service = (BudgetOrderService)user.GetService(AdWordsService.v201802.BudgetOrderService);
service.RequestHeader = new FixedRequestHeader(service.RequestHeader);

// use service as usual
```

**Comment by tonmanna** [comment]
Big thanks @pieceofsummer  for your share solution. I submit discussion on Adwords Group They need track our SOAP Log. Your expect is correct is related to Google AdWords server because we also have this problem with any API on this morning.

**Comment by AnashOommen** [comment]
Folks, I don't think this is client-library-specific issue, we are investigating this at our end as a possible server-side issue.

**Comment by AnashOommen** [comment]
Pls follow https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/adwords-api/Zx3B8xO5rW4/DeSDQKo-BgAJ

**Comment by pieceofsummer** [comment]
@AnashOommen we don't actually think it is a client issue either, because it was working fine until a few days ago, and then just broke without any changes at the client side :)

The code above is just a temporary client-side workaround I came up with for this server issue. It affects our business-critical workflows and it was essential to fix it ASAP.

**Comment by AnashOommen** [comment]
Got it, and thanks for the suggested workaround, that is helping us in our investigation.

**Comment by pieceofsummer** [comment]
Seems to be fixed now

**Comment by dgaleanopmx** [comment]
This happened using the CustomerSyncService on my end.
The temporary fix has resolved it.
It is not a client library issue because this has been working fine since February.
Google made some change on their servers...

**Comment by tonmanna** [comment]
@dgaleanopmx  You still need to add CustomHeader config? We don't need to customer header anymore sine 4 days ago.

**Comment by dgaleanopmx** [comment]
@tonmanna is working without it.
Weird, started to get this since Monday but I think they reverted the changes.

**Comment by christopherseeley** [comment]
Fixed in v23.3.0
